bot/exts/moderation/voice_gate.py

```diff
@@ -0,0 +1,136 @@
+import logging
+from contextlib import suppress
+from datetime import datetime, timedelta
+
+import discord
+from dateutil import parser
+from discord.ext.commands import Cog, Context, command
+
+from bot.api import ResponseCodeError
+from bot.bot import Bot
+from bot.constants import Channels, Event, MODERATION_ROLES, Roles, VoiceGate as VoiceGateConf
+from bot.decorators import has_no_roles, in_whitelist
+from bot.exts.moderation.modlog import ModLog
+from bot.utils.checks import InWhitelistCheckFailure
+
+log = logging.getLogger(__name__)
+
+# Messages for case when user don't meet with requirements
+NOT_ENOUGH_MESSAGES = f"haven't sent at least {VoiceGateConf.minimum_messages} messages"
+NOT_ENOUGH_DAYS_AFTER_VERIFICATION = f"haven't been verified for at least {VoiceGateConf.minimum_days_verified} days"
+VOICE_BANNED = "are voice banned"
+
+FAILED_MESSAGE = """{user} you don't meet with our current requirements to pass Voice Gate. You {reasons}."""
+
+
+class VoiceGate(Cog):
+    """Voice channels verification management."""
+
+    def __init__(self, bot: Bot):
+        self.bot = bot
+
+    @property
+    def mod_log(self) -> ModLog:
+        """Get the currently loaded ModLog cog instance."""
+        return self.bot.get_cog("ModLog")
+
+    @command(aliases=('voiceverify', 'vverify', 'voicev', 'vv'))
+    @has_no_roles(Roles.voice_verified)
+    @in_whitelist(channels=(Channels.voice_gate,), redirect=None)
+    async def voice_verify(self, ctx: Context, *_) -> None:
+        """
+        Apply to be able to use voice within the Discord server.
+
+        In order to use voice you must meet all three of the following criteria:
+        - You must have over a certain number of messages within the Discord server
+        - You must have accepted our rules over a certain number of days ago
+        - You must not be actively banned from using our voice channels
+        """
+        try:
+            data = await self.bot.api_client.get(f"bot/users/{ctx.author.id}/metricity_data")
+        except ResponseCodeError as e:
+            if e.status == 404:
+                await ctx.send(f":x: {ctx.author.mention} Unable to find Metricity data about you.")
+                log.info(f"Unable to find Metricity data about {ctx.author} ({ctx.author.id})")
+            else:
+                log.warning(f"Got response code {e.status} while trying to get {ctx.author.id} metricity data.")
+                await ctx.send(":x: Got unexpected response from site. Please let us know about this.")
+            return
+
+        # Pre-parse this for better code style
+        data["verified_at"] = parser.isoparse(data["verified_at"])
```

Comment:
One thing I need to mention which I forgot. This might be null. We need to account for that, I'm tempted to say we give away a free pass when this happens and just set it to a time delta 3 days ago. The data will correct itself after a week or so and set the users to have a valid date.